[AArch64] New subtarget features to control ldp and stp formation, fo…

manosanaggh opened the pull request:
…cused on ampere1 and ampere1a.

On some AArch64 cores, including Ampere's ampere1 and ampere1a architectures, load and store pair instructions are faster compared to simple loads/stores only when the alignment of the pair is at least twice that of the individual element being loaded.

Based on that, this patch introduces four new subtarget features, two for controlling ldp and two for controlling stp, to cover the ampere1 and ampere1a alignment needs and to enable optional fine-grained control over ldp and stp generation in general. The latter can be utilized by another cpu, if there are possible benefits
with a different policy than the default provided by the compiler.

More specifically, for each of the ldp and stp respectively we have:

- disable-ldp/disable-stp: Do not emit ldp/stp.
- ldp-aligned-only/stp-aligned-only: Emit ldp/stp only if the source pointer is aligned to at least double the alignment of the type.

Therefore, for -mcpu=ampere1 and -mcpu=ampere1a
ldp-aligned-only/stp-aligned-only become the defaults, because of the benefit from the alignment, whereas for the rest of the cpus the default behaviour of the compiler is maintained.

manosanaggh commented:
This is a new version, which was requested here: https://reviews.llvm.org/D159480

davemgreen reviewed (commented):
Thanks for moving this over. It looks sensible to me.

manosanaggh commented on llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:
```diff
@@ -2136,6 +2136,18 @@ bool AArch64LoadStoreOpt::tryToPairLdStInst(MachineBasicBlock::iterator &MBBI) {
 
+  // Fetch the memoperand of the load/store that is a candidate for
+  // combination.
+  MachineMemOperand *memOp = MBBI->memoperands().data()[0];
```
Well, I thought that that way I could get the memoperand directly in a single line of code without needing to ask if its a load or store and have to choose between the first or the second one (general operand) respectively.

manosanaggh commented on llvm/test/CodeGen/AArch64/disable-ldp-ampere1.ll:
```diff
@@ -0,0 +1,113 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 3
+; RUN: llc < %s -O2 -mtriple=aarch64 -mcpu=ampere1 -mattr=+disable-ldp | FileCheck %s
```
Yes sure, I was kind of confused with that. Thanks for the tip!

manosanaggh commented:
> Thanks for moving this over. It looks sensible to me.

Thank you too for the tips! I am wondering if these x64 regressions in the checks are a thing or false positives, but I am going to have a real quick test tomorrow in a x64 machine.

manosanaggh commented:
> > Thanks for moving this over. It looks sensible to me.
> 
> Thank you too for the tips! I am wondering if these x64 regressions in the checks are a thing or false positives, but I am going to have a real quick test tomorrow in a x64 machine.

I have to switch to debug build for that on x64 for that. Should be a test-issue.

davemgreen commented on llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:
```diff
@@ -2159,6 +2171,24 @@ bool AArch64LoadStoreOpt::tryToPairLdStInst(MachineBasicBlock::iterator &MBBI) {
+
+    // Get the needed alignments to check them if
+    // ldp-aligned-only/stp-aligned-only features are opted.
+    uint64_t MemAlignment = MemOp->getAlign().value();
```
This looks like it could be performed higher, up above the call to findMatchingInsn? So long as it doesn't need to check the second instruction in the pair too.

If so there are some other conditions checked in isCandidateToMergeOrPair. I'm not sure why they are separated out, but it might be good to keep those kinds of checks together in the same function.

manosanaggh replied:
I think I will move it right under the call to isCandidateToMergeOrPair and not inside it, because I do not want to alter the outcome of the function for merges (merge != pair).

manosanaggh replied:
This should remain where it is, because it breaks in case "if (Paired != E)" is not entered and the function is going to "return false".

manosanaggh commented on llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:
```diff
@@ -2136,6 +2136,18 @@ bool AArch64LoadStoreOpt::tryToPairLdStInst(MachineBasicBlock::iterator &MBBI) {
 
+  // Fetch the memoperand of the load/store that is a candidate for
+  // combination.
+  MachineMemOperand *MemOp = MI.memoperands().front();
```
This may have to do with the regression on x64, because its a memoperand assertion. Although I am still confused a bit. Thanks for the hint! Should help a lot!

manosanaggh commented on llvm/test/CodeGen/AArch64/ldp-stp-control-features.ll:
```diff
@@ -0,0 +1,390 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 3
+; RUN: llc < %s -O2 -mtriple=aarch64 -mcpu=ampere1 | FileCheck %s --check-prefixes=CHECK-LABEL,CHECK,CHECK-NEXT
```
Alright!

manosanaggh commented:
The regression on my test case, caught by the ci, also happens with the upstream llvm (main branch) and has nothing to do with my changes. MachineScheduler pass causes it. I may have to file a report on it.

manosanaggh commented:
Filed a report for the crash on: https://github.com/llvm/llvm-project/issues/66328

davemgreen commented on llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:
```diff
@@ -2159,6 +2167,30 @@ bool AArch64LoadStoreOpt::tryToPairLdStInst(MachineBasicBlock::iterator &MBBI) {
+
+    // Fetch the memoperand of the load/store that is a candidate for
+    // combination.
+    MachineMemOperand *MemOp = MI.memoperands().front();
```
I think front() might assert if the array is empty. Can this use `MI.memoperands_empty() ? nullptr : MI.memoperands().front()`?

manosanaggh replied:
Sure, pushing real soon.

davemgreen commented on llvm/test/CodeGen/AArch64/ldp-stp-control-features.ll:
```diff
@@ -0,0 +1,390 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 3
+; RUN: llc < %s -O2 -mtriple=aarch64 -mcpu=ampere1 | FileCheck %s --check-prefixes=CHECK
+; RUN: llc < %s -O2 -mtriple=aarch64 -mcpu=ampere1a | FileCheck %s --check-prefixes=CHECK
+; RUN: llc < %s -O2 -mtriple=aarch64 | FileCheck %s --check-prefixes=CHECK-LABEL-DEFAULT,CHECK-DEFAULT,CHECK-NEXT-DEFAULT
```
These should be --check-prefix=CHECK-DEFAULT, same for the ones below. The LABEL and NEXT prefixes are automatic.

They should generate check lines like `; CHECK-DEFAULT-NEXT:`, I believe.

manosanaggh replied:
Oh ok, I thought this was done with the normal check only. Good to know from now on! Thanks!

davemgreen reviewed (commented):
Sorry I missed that the check lines are still a bit off. If you fix that then this LGTM. Thanks

manosanaggh committed:
[AArch64] New subtarget features to control ldp and stp formation, focused on ampere1 and ampere1a.

On some AArch64 cores, including Ampere's ampere1 and ampere1a
architectures, load and store pair instructions are faster compared
to simple loads/stores only when the alignment of the pair is at least
twice that of the individual element being loaded.

Based on that, this patch introduces four new subtarget features,
two for controlling ldp and two for controlling stp, to cover
the ampere1 and ampere1a alignment needs and to enable optional
fine-grained control over ldp and stp generation in general.
The latter can be utilized by another cpu, if there are possible
benefits
with a different policy than the default provided by the compiler.

More specifically, for each of the ldp and stp respectively we have:

- disable-ldp/disable-stp: Do not emit ldp/stp.
- ldp-aligned-only/stp-aligned-only: Emit ldp/stp only if the source
pointer is aligned to at least double the alignment of the type.

Therefore, for -mcpu=ampere1 and -mcpu=ampere1a
ldp-aligned-only/stp-aligned-only become the defaults because,
of the benefit from the alignment, whereas for the rest
of the cpus the default behaviour of the compiler is maintained.

davemgreen reviewed (approved):
Thanks LGTM

manosanaggh commented:
> Thanks LGTM

Thank you too!

davemgreen commented:
There is hopefully a fix for the issues with the ampere1 model in #66384

@manosanaggh I didn't seem to be able to add you as a reviewer, but it would be good if you could take a look. Thanks
